pkg/tuned/cmd/render/render.go

```diff
@@ -0,0 +1,251 @@
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog"
+)
+
+var (
+	manifestScheme = runtime.NewScheme()
+	codecFactory   serializer.CodecFactory
+	runtimeDecoder runtime.Decoder
+)
+
+func init() {
+	utilruntime.Must(performancev2.AddToScheme(manifestScheme))
+	utilruntime.Must(mcfgv1.Install(manifestScheme))
+	codecFactory = serializer.NewCodecFactory(manifestScheme)
+	runtimeDecoder = codecFactory.UniversalDecoder(
+		performancev2.GroupVersion,
+		mcfgv1.GroupVersion,
+		tunedv1.SchemeGroupVersion,
+	)
+}
+
+func render(inputDir, outputDir string) error {
+	klog.Info("Rendering files from: ", inputDir)
+	klog.Info("Rendering files into: ", outputDir)
+
+	bootstrapSafeEnv := os.Getenv("CLUSTER_NODE_TUNED_BOOTSTRAP_SAFE_ENV")
+	if len(bootstrapSafeEnv) == 0 {
+		return fmt.Errorf("Should only be run on bootstrap safe environment. Please define env var 'CLUSTER_NODE_TUNED_BOOTSTRAP_SAFE_ENV' ")
+	}
+
+	// Get pools, mConfigs and profile from inputDir
+	// Read asset directory fileInfo
+	filePaths, err := util.ListFiles(inputDir)
+	klog.V(4).Infof("listed files: %v", filePaths)
+	if err != nil {
+		return err
+	}
+	// Make output dir if not present
+	err = os.MkdirAll(outputDir, os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	var (
+		perfProfiles []*performancev2.PerformanceProfile
+		mcPools      []*mcfgv1.MachineConfigPool
+		mcConfigs    []*mcfgv1.MachineConfig
+		tuneD        []*tunedv1.Tuned
+	)
+
+	// Iterate through the file paths and read in desired files
+	for _, path := range filePaths {
+		file, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("error opening %s: %w", file.Name(), err)
+		}
+		defer file.Close()
+
+		manifests, err := util.ParseManifests(file.Name(), file)
+		if err != nil {
+			return fmt.Errorf("error parsing manifests from %s: %w", file.Name(), err)
+		}
+
+		// Decode manifest files
+		for idx, m := range manifests {
+			obji, err := runtime.Decode(runtimeDecoder, m.Raw)
+			if err != nil {
+				if runtime.IsNotRegisteredError(err) {
+					klog.V(4).Infof("skipping path %q [%d] manifest because it is not part of expected api group: %v", file.Name(), idx+1, err)
+					continue
+				}
+				return fmt.Errorf("error parsing %q [%d] manifest: %w", file.Name(), idx+1, err)
+			}
+
+			switch obj := obji.(type) {
+			case *performancev2.PerformanceProfile:
+				perfProfiles = append(perfProfiles, obj)
+			case *mcfgv1.MachineConfigPool:
+				mcPools = append(mcPools, obj)
+			case *mcfgv1.MachineConfig:
+				mcConfigs = append(mcConfigs, obj)
+			case *tunedv1.Tuned:
+				tuneD = append(tuneD, obj)
+			default:
+				klog.Infof("skipping %q [%d] manifest because of unhandled %T", file.Name(), idx+1, obji)
+			}
+		}
+	}
+
+	if len(perfProfiles) != 1 {
```

Comment:
Are you sure about this? It would work for SNO, but can this logic help with bigger clusters too? There can be more PerfProfiles with associated MCPs I think.